### ISSUE TITLE
Running tests with compile: give hand if there's an ipdb breakpoint (feature proposal)

### ISSUE THREAD

**Issue by vindarel** [jorgenschaefer/elpy#932]
Hi !

So elpy's test commands uses `M-x compile`. The thing is, for development I put ipdb breakpoints inside of my tests.

If I run `elpy-test-django-runner`, it uses a compilation buffer, the test stops at the breakpoint but by default I'm not given the hand into the ipdb prompt.

I've been using Mickeyp's solution with great success: https://masteringemacs.org/article/compiling-running-scripts-emacs We just have to defadvice compile.

Looks like it would be a good additon to elpy to me. What do you think ?

**Comment by marsam** [comment]
FWIW, I have a `compilation-filter-hook` which automatically detects a pdb prompts and enables `inferior-python-mode` from python.el : https://github.com/emacs-pe/python-test.el/blob/c478bb0/python-test.el#L320-L330

**Comment by jorgenschaefer** [comment]
Oh, that's a nice idea, thank you for the suggestion and the code there! :-)

**Comment by vindarel** [comment]
Cool, I'd be happy to help on that :)

@marsam Thanks for the link. Does that do the same thing as I described ? (what is "do something" ?)

**Comment by marsam** [comment]
@vindarel yep, does the same, but is a hack that enables `inferior-python-mode` instead of `comint-mode`, the advantage is that `inferior-python-mode` offers pdb tracking. Thanks I realized how silly is that doctring, I'll fix it later.

**Comment by humitos** [comment]
I wasn't able to setup your configuration. I pasted the elisp code into my config file but for some reason the "*compilation*" buffer is still read only.

Once I'm in the compilation buffer I ran manually:

    (setq buffer-read-only nil)
    (inferior-python-mode)

and it works but for some reason it's not detecting the "(Pdb) " prompt automatically.

Do you know how I can debug this?

**Comment by vindarel** [comment]
Hi Humitos,
maybe it is because your breakpoint is in another file that the one you call compile from (just ran into that).

We could enable comint-mode everytime:

```lisp
(defadvice compile
    (before ad-compile-smart activate)
  "Advises `compile' so it sets the argument COMINT to t if breakpoints are present in `python-mode' files"
  (when (derived-mode-p major-mode 'python-mode)
      (progn
        (ad-set-arg 1 t)
        (switch-to-buffer-other-window "*compilation*"))))
```

(Do you see a drawback in doing so ?)

**Comment by marsam** [comment]
@humitos sorry for the late response. If you mean my configuration you can try the following: Add [`python-test.el`](https://github.com/emacs-pe/python-test.el/blob/master/python-test.el) to your `load-path` and add the following to your `init.el`:

    (add-hook 'compilation-filter-hook 'python-test-track-pdb-prompt)

Feel free to [create an issue in python-test.el](https://github.com/emacs-pe/python-test.el/issues) if it doesn't work

**Comment by excalamus** [comment]
This issue is independent of Elpy, so far as I can tell.  When Elpy calls pytest with the "-s" argument, control is given to (i)pdb.  It is `compile`/the `*compilation*` buffer which is the problem.

@humitos is on the right track.  The compilation buffer is read-only by default.  This is disabled via `(setq buffer-read-only nil)`.  The compilation buffer also needs to be in `comint-mode` in order to interact with (i)pdb.  `inferior-python-mode` is derived from `comint-mode`.  So, those are the right ingredients.

The difficulty is how to use them.  `compile`/the compilation buffer are unruly.

The approach by @marsam  works.  To my understanding, it does so intercepting the comint process and checking for the `python-shell-prompt-pdb-regexp` (included with `python.el`).  The regexp is  `"[(<]*[Ii]?[Pp]db[>)]+ "`.  When the regxep is present, `python-test-track-pdb-prompt` sets the buffer to be writable and enables `inferior-python-mode`.  I think it does this on every comint insertion (e.g. every time you step using the debugger).

```python
(defun python-test-track-pdb-prompt ()
  "Do something if pdb prompt is detected."
  (let ((output (buffer-substring-no-properties compilation-filter-start (point-max))))
    (when (string-match-p python-shell-prompt-pdb-regexp output)
      (message "Entering pdb...")
      (setq buffer-read-only nil)
      (let ((python-shell--interpreter nil)
            (python-shell--interpreter-args nil))
        (set-process-filter (get-buffer-process (current-buffer)) 'comint-output-filter)
        (inferior-python-mode)
        (run-hook-with-args 'comint-output-filter-functions output)))))
```

The method linked by @vindarel is a little more efficient, although it will not work because it is outdated.  It overrides the `compile` function itself to first search the current buffer for the string `import pdb; pdb.set_trace() ## DEBUG ##`.  When found, it calls `compile` with its second argument true (i.e. to make the compilation buffer be in Comint mode with `compilation-shell-minor-mode`).  Unfortunately, it uses the "old" Emacs advice functionality (`ad-set-arg`) which will not work on current Emacs,

```lisp
;; taken from https://masteringemacs.org/article/compiling-running-scripts-emacs
(require 'python)
(defun python--add-debug-highlight ()
  "Adds a highlighter for use by `python--pdb-breakpoint-string'"
  (highlight-lines-matching-regexp "## DEBUG ##\\s-*$" 'hi-red-b))

(add-hook 'python-mode-hook 'python--add-debug-highlight)

(defvar python--pdb-breakpoint-string "import pdb; pdb.set_trace() ## DEBUG ##"
  "Python breakpoint string used by `python-insert-breakpoint'")

(defun python-insert-breakpoint ()
  "Inserts a python breakpoint using `pdb'"
  (interactive)
  (back-to-indentation)
  ;; this preserves the correct indentation in case the line above
  ;; point is a nested block
  (split-line)
  (insert python--pdb-breakpoint-string))
(define-key python-mode-map (kbd "<f5>") 'python-insert-breakpoint)

(defadvice compile (before ad-compile-smart activate)
  "Advises `compile' so it sets the argument COMINT to t
if breakpoints are present in `python-mode' files"
  (when (derived-mode-p major-mode 'python-mode)
    (save-excursion
      (save-match-data
        (goto-char (point-min))
        (if (re-search-forward (concat "^\\s-*" python--pdb-breakpoint-string "$")
                               (point-max) t)
            ;; set COMINT argument to `t'.
            (ad-set-arg 1 t))))))
```

I like the advising approach, personally, as it only parses a buffer once.  

tl;dr
=====
Here is what I recommend:

```lisp
;; Have pytest capture sdtin
(setq elpy-test-pytest-runner-command '("py.test" "-s"))

;; Match on any appearance of pdb or ipdb
(setq my-python-break-regexp "[i]?pdb")

;; Use the new Emacs advice to search the current buffer for pdb/ipdb only if current buffer is derived from python-mode.  Call compile with comint arg t if found.
(defun my-compile-advice (args)
  "Advises `compile' so it sets the argument COMINT to t
if breakpoints are present in `python-mode' files"
  (when (derived-mode-p major-mode 'python-mode)
    (save-excursion
      (save-match-data
        (goto-char (point-min))
        (if (re-search-forward my-python-break-regexp (point-max) t)
            ;; set COMINT argument to `t'.
            (let ((command (car args)))
              (setq args (list command t)))))))
  args)

(advice-add 'compile :filter-args #'my-compile-advice)
```

You will also likely find that the compilation buffer window is annoying and difficult to tame.  It pops up all over and doesn't like to remain put.  This code forces it to stay in one window:

```lisp
;; black magic taken from https://emacs.stackexchange.com/a/338/15177
(add-to-list 'display-buffer-alist
         '("^\\*compilation\\*". ((display-buffer-reuse-window)  .
                                  ((reusable-frames . t)
                                  (inhibit-same-window . t)))))
```

Hopefully that helps!
